docs/_specs/application/application.md

```diff
@@ -0,0 +1,92 @@
+
+### Identities & Access Controls - Users & Groups
+
+It hasn’t be decided whether or not we will manage users, groups and access controls within dispatch or with the help of an existing framework. If we decide we should include them as dispatch managed entities, they should be system-scoped. In addition to it, there should be some description/information on which users/groups have permissions on which applications.
+
+### Functions, Images & Secrets
+
+In our vision of Dispatch, functions, images and secrets are more like packages, libraries or source code of a traditional application. They are not the runtime resources of a single Application instance. With this point of perspective, we think there are potential use cases that a function or function image could be used by multiple applications. However, note there indeed are some debates going on here. Rather than categorizing them deterministically into one single basket, we decide to keep it flexible in implementation, and iterate quickly as we see sufficient use cases in the future.
+
+Currently, we will temporarily put them into the one-to-many category. However, again, we will be flexible and extensible in implementation detail.
+
+
+### Function Runs
+
+Function Runs are entities generated from function executations. They are runtime entities and they should belongs to a single Application.
+
+### APIs
+
+APIs are considered tightly bound with an Application. Different applications have no reason to share one API. Therefore, one API should stay in the same Application as its bounded Function.
+
+### Event Drivers, Topics & Subscriptions
+
+There are indeed some use cases that some events maybe shared by multiple applications, (e.g. events from the dispatch or kubernetes internal, VM lifecycle events may be in demand for several applications).
+
+However, in some other use cases, events maybe only be used by specific applications, e.g. a PR build failure event triggers a report to a internal Slack channel, as we can image, it will almost be used by only one dispatch application.
+
+Both shared one-to-many and many-to-many events are valid use cases. When there’s not enough real use cases it is hard to decide which is dominant and which is rare. Again, we decide to keep it flexible at this point. In the first iteration, we will set them as private to an Application for now.
+
+As a side note, although Event Topics are not a dispatch entity now, we should also consider to scope them by Application (especially if we decide that Event Subscriptions should be scoped). For example, add the Application name as a prefix of a topic name.
+
+## Implementation Detail
+
+
+### The Application Entity
+
+As most entities are organized by applications, an Application itself also has a life cycle. To manage applications, it is best to manage and maintain it with the existing Entity Store.
```

Comment:
What happens to entities which reference applications when applications are deleted?  How does the application lifecycle impact the entities which reference the application?

Reply:
that's a good question, with many-to-many relationship, I can image we should not delete associated entities if one application is deleted. 

we should also probably have a DeleteOption say with DeleteOption = "cascaded", delete all associated entities, with DeleteOption="regular", only the application entity will be delete.

Reply:
I think we should cascade in this scenario.  Also for APIs.  They should probably be cascaded when a function is deleted.

Reply:
got you, we could have the default behavior as cascaded for now.

Reply:
Kubernetes has an option, enabled by default, to cascade delete the resources when deleting a namespace. We could follow this approach (delete by default).